src/corelibs/U2Formats/src/BAMUtils.cpp

```diff
@@ -606,27 +631,48 @@ static int bamMergeCore(const QString& outFileName, const QList<QString>& filesT
     return 0;
 }
 
-GUrl BAMUtils::mergeBam(const QStringList& bamUrls, const QString& mergedBamTargetUrl, U2OpStatus& os) {
-    coreLog.details(tr(R"(Merging BAM files: "%1". Resulting merged file is: "%2")")
-                        .arg(QString(bamUrls.join(",")))
-                        .arg(QString(mergedBamTargetUrl)));
+static constexpr int MAX_FILES_OPENED = 100;
+static int recursiveBamMergeCore(const QString& outFileName, const QList<QString>& filesToMerge, QStringList& intermediateFiles) {
+    int size = filesToMerge.size();
+    CHECK(size != 0, -1);
 
-    int rc = bamMergeCore(mergedBamTargetUrl, bamUrls);
-    CHECK_EXT(rc >= 0, os.setError(tr("Failed to merge BAM files: %1 into %2").arg(bamUrls.join(",")).arg(mergedBamTargetUrl)), {});
-    return mergedBamTargetUrl;
+    auto mergeSplit = U2Region::split({ 0, size }, MAX_FILES_OPENED);
+    if (mergeSplit.size() == 1) {
+        return localBamMergeCore(outFileName, filesToMerge);
+    }
+
+    QStringList newOutFileNameList;
+    for (int i = 0; i < mergeSplit.size(); i++) {
+        const auto& currentRange = mergeSplit.at(i);
+         QList<QString> newFilesToMerge = filesToMerge.mid(currentRange.startPos, currentRange.endPos());
+        QString newOutFileName = newFilesToMerge.first();
+        do {
+            // Remove ".bam" from the end
+            newOutFileName = newOutFileName.left(newOutFileName.size() - 4);
+            newOutFileName = newOutFileName + "_" + QString::number(i) + ".bam";
+        } while (QFile(newOutFileName).exists());
```

Comment:
nit: access to the FS is not free. A better & simpler 1 liner solution is possible here: append a randomly generated uuid() to the original file name.

Reply:
OK